community: Update get_table_info() for better explainability

manuelrech opened the pull request:
I propose this to have both custom information, info from the data def lang and example rows.

I have noticed that when you provide the dictionary with custom information about your tables you do not get the data definition language (the schema with create table ...) and the rows examples. I modified just a couple rows to accommodate for CUSTOM DESCRIPTIONS + DEFAULT DESCRIPTIONS

@RechManuel my X account

manuelrech committed:
Update get_table_info() to have both custom information, info from the data def lang and example rows

I have notices that when you provide the dictionary with custom information about your tables you do not get the data definition language (the schema with create table ...) and the rows examples. I modified just a couple rows to accommodate for CUSTOM DESCRIPTIONS + DEFAULT DESCRIPTIONS

eyurtsev commented:
1) Could you provide a before and after from this function? Otherwise it's not clear whether default behavior should be changed for all users

If we decide to incorporate change, we'd need to:
2) Add a unit test for this functionality
3) Potentially provide users with opt-in control

manuelrech committed:
Merge branch 'master' into patch-1

manuelrech committed:
Merge branch 'master' into patch-1

manuelrech committed:
added opt in control in get_table_info()

added opt-in control for users in get_table_info() so that they can decide whether to use custom information AND the default (schema and sample rows) or only the custom information as it was before

manuelrech commented:
LC
The last part of the function `get_table_info()` was
```python
for table in meta_tables:
            if self._custom_table_info and table.name in self._custom_table_info:
                tables.append(self._custom_table_info[table.name])
                continue

            # Ignore JSON datatyped columns
            for k, v in table.columns.items():
                if type(v.type) is NullType:
                    table._columns.remove(v)

            # add create table command
            create_table = str(CreateTable(table).compile(self._engine))
            table_info = f"{create_table.rstrip()}"
            has_extra_info = (
                self._indexes_in_table_info or self._sample_rows_in_table_info
            )
            if has_extra_info:
                table_info += "\n\n/*"
            if self._indexes_in_table_info:
                table_info += f"\n{self._get_table_indexes(table)}\n"
            if self._sample_rows_in_table_info:
                table_info += f"\n{self._get_sample_rows(table)}\n"
            if has_extra_info:
                table_info += "*/"
            tables.append(table_info)
        tables.sort()
        final_str = "\n\n".join(tables)
        return final_str
```
and I changed into 

```python
for table in meta_tables:
            if self._custom_table_info and table.name in self._custom_table_info:
                if self._custom_and_default_info:
                    table_info = f"{table} description: {self._custom_table_info[table.name]}\n\n"
                else:
                    tables.append(self._custom_table_info[table.name])
                    continue

            # Ignore JSON datatyped columns
            for k, v in table.columns.items():
                if type(v.type) is NullType:
                    table._columns.remove(v)

            # add create table command
            create_table = str(CreateTable(table).compile(self._engine))
            table_info =  table_info + f"{table} DDL:{create_table.rstrip()}" if self._custom_and_default_info else f"{table} DDL:{create_table.rstrip()}"
            has_extra_info = (
                self._indexes_in_table_info or self._sample_rows_in_table_info
            )
            if has_extra_info:
                table_info += "\n\n/*"
            if self._indexes_in_table_info:
                table_info += f"\n{self._get_table_indexes(table)}\n"
            if self._sample_rows_in_table_info:
                table_info += f"\n{self._get_sample_rows(table)}\n"
            if has_extra_info:
                table_info += "*/"
            tables.append(table_info)
        tables.sort()
        final_str = "\n\n".join(tables)
        return final_str
```

I added the opt-in control as you suggested me but I am sorry I do not know how to create proper unit test, I am not very experienced, I just added this feature locally and I thought of sharing it!

eyurtsev commented:
@manuelrech thanks! I meant the before and after on the return value of the function -- since It's difficult to tell what the change is doing. the diff on the function is visible on the PR itself, it's a bit hard to tell how the change affects downstream code etc

manuelrech commented:
Okay I see. So supposing you instantiate you `SQLDatabase` like this

```python
db = SQLDatabase(
    engine=engine,
    ignore_tables=['team_data_short', 'player_data_short', 'kpi_definitions_short', 'affiliations'],
    custom_table_info={
        "MatchStats": "Stores match setups and outcomes, featuring dual entries per match ID to represent both competing teams. Each record uniquely identified by a composite primary key consisting of match_id and team_id.",
        "PlayerStats": "Details player-specific performance metrics, akin to MatchStats' structure. Unique entries identified by a composite primary key (match_id, player_id), with match_id linking to MatchStats and team_id used for additional relational context but not as a part of the primary key.", 
        "TeamStats": "Captures team performance statistics, structured similarly to MatchStats. Each entry is uniquely identified by a composite primary key (match_id, team_id), which also serves as a foreign key referencing MatchStats"
        }
    )
```
for example for the `SQLDatabaseToolkit()` to be used in the react agent for querying a db. 

Given a question, the agent will look at the table names and in order to generate the sql query(s). To answer it need to look at the schemas of some of the tables he thinks are relevant. The problem comes here. In the `for` loop shown above over the meta_tables, the code will get a custom description (as the one I have provided) and it will leave the current iteration of the meta_tables on the `continue` keyword, not adding the DDL as in `create_table = str(CreateTable(table).compile(self._engine))` and the sample rows as in `table_info += f"\n{self._get_sample_rows(table)}\n"`, which are both fundamental for generating a good query. 

As an example, before I would get 

```
Tool call: sql_db_schema
    Tool arguments: {'table_names': 'PlayerStats'}
----------------------------------------------------------------------------------------------------
Details player-specific performance metrics, akin to MatchStats' structure. Unique entries identified by a composite primary key (match_id, player_id), with match_id linking to MatchStats and team_id used for additional relational context but not as a part of the primary key.
```

and now I get
```
Tool call: sql_db_schema
    Tool arguments: {'table_names': 'PlayerStats'}
----------------------------------------------------------------------------------------------------
Tool response: PlayerStats description: Details player-specific performance metrics, akin to MatchStats' structure. Unique entries identified by a composite primary key (match_id, player_id), with match_id linking to MatchStats and team_id used for additional relational context but not as a part of the primary key.

PlayerStats DDL:
CREATE TABLE `PlayerStats` (
	match_id INTEGER NOT NULL, 
	player_id INTEGER NOT NULL, 
	team_id INTEGER, 
	`role` TEXT COMMENT 'can be Goalkeeper, Defender, Midfielder, Forward', 
	start11 TINYINT COMMENT '1 if the player was in the starting11', 
	minutes DOUBLE COMMENT 'minutes played by the player per game', 
	`teamActions` INTEGER COMMENT 'number of team actions while player was on field', 
	`oppTeamActions` INTEGER COMMENT 'number of oppenent''s team actions while player was on field', 
	`timeIn` INTEGER COMMENT 'second in which the player entered the game, 0 if he/she was on the starting 11', 
	`timeOut` INTEGER COMMENT 'second in which the player left the game, 0 for the substituted in and max value for those who played the entire game', 
	`normalizedGame` DOUBLE COMMENT 'normalization constant representing how much of the total gametime was played', 
	`gamePlayed` TINYINT COMMENT '1 if the player played in the game', 
	`substitutedOff` TINYINT COMMENT '1 if the player was substituted off', 
	`substitutedOn` TINYINT COMMENT '1 if the player was substituted on', 
	passes INTEGER COMMENT 'number of player passes', 
	`successfulPasses` INTEGER COMMENT 'number of player''s successful passes', 
	`receivedPasses` INTEGER COMMENT 'number of player''s received passes', 
	`forwardPasses` INTEGER COMMENT 'number of player''s passes to forwards', 
	crosses INTEGER COMMENT 'number of player''s crosses ', 
	`successfulCrosses` INTEGER COMMENT 'number of successful crosses made', 
	`keyPasses` INTEGER COMMENT 'number of player''s key passes', 
	assists INTEGER COMMENT 'number of player''s assists', 
	`targetShots` INTEGER COMMENT 'number of player''s shots on target', 
	`postShots` INTEGER COMMENT 'number of player''s shots on post', 
	`boxGoals` INTEGER COMMENT 'number of player''s goals from the box', 
	`outBoxGoals` INTEGER COMMENT 'number of player''s goals from outside the box', 
	shots INTEGER COMMENT 'amount of player''s shots taken', 
	`boxShots` INTEGER COMMENT 'number of player''s shots from inside the box.', 
	`outBoxShots` INTEGER COMMENT 'number of player''s shots from outside the box', 
	goals INTEGER COMMENT 'number of goals made', 
	PRIMARY KEY (match_id, player_id), 
	CONSTRAINT fk_playerstats_match FOREIGN KEY(match_id, team_id) REFERENCES `MatchStats` (match_id, team_id)
)DEFAULT CHARSET=utf8mb4 ENGINE=InnoDB COLLATE utf8mb4_0900_ai_ci

[/](https://file+.vscode-resource.vscode-cdn.net/)*
3 rows from PlayerStats table:
match_id	player_id	team_id	role	start11	minutes	teamActions	oppTeamActions	timeIn	timeOut	normalizedGame	gamePlayed	substitutedOff	substitutedOn	passes	successfulPasses	receivedPasses	forwardPasses	crosses	successfulCrosses	keyPasses	assists	targetShots	postShots	boxGoals	outBoxGoals	shots	boxShots	outBoxShots	goals
CANNOT DISCLOSE THE CONTENTS FOR PRIVACY REASONS BUT THEY WOULD BE THERE*/

```

manuelrech committed:
Merge branch 'langchain-ai:master' into patch-1

manuelrech committed:
Merge branch 'master' into patch-1

manuelrech commented:
Also, I think this should be a default behaviour because most likely users will provide a string description and not the DDL and sample rows, so the default flag could be set as `True` (now is `False`)

eyurtsev commented on libs/community/langchain_community/utilities/sql_database.py:
```diff
@@ -59,6 +59,7 @@ def __init__(
         view_support: bool = False,
         max_string_length: int = 300,
         lazy_table_reflection: bool = False,
+        custom_and_default_info: bool = False,
```
This variable is not self explanatory. Could you add documentation for this?

A good name for this variable would include a prefix of "include_"

For example, "include_table_schema_in_description" or something in that vain

eyurtsev replied:
Not sure if you're willing to tackle, but we'd benefit from general help improving documentation for this class (i.e., documenting the parameters + usage)

The class should provide functionality to retriever some raw information and allow the user to compose into an arbitrary prompt. It's OK that we have some default strategy for composition, but it's not flexible enough if users can't access the raw data

eyurtsev commented:
Feel free to re-open if you can address comments
